src/ecredis.erl

```diff
@@ -0,0 +1,420 @@
+            {ok, ClusterInfo} ->
+                ClusterInfo;
+            _ ->
+                get_cluster_slots(ClusterName, T)
+          end;
+        _ ->
+            get_cluster_slots(ClusterName, T)
+  end.
+
+
+-spec get_cluster_slots_from_single_node(#node{}) -> [[bitstring() | [bitstring()]]].
+get_cluster_slots_from_single_node(Node) ->
+    [[<<"0">>, integer_to_binary(?REDIS_CLUSTER_HASH_SLOTS-1),
+    [list_to_binary(Node#node.address), integer_to_binary(Node#node.port)]]].
+
+
+-spec parse_cluster_slots([[bitstring() | [bitstring()]]]) -> [#slots_map{}].
+parse_cluster_slots(ClusterInfo) ->
+    parse_cluster_slots(ClusterInfo, 1, []).
+
+parse_cluster_slots([[StartSlot, EndSlot | [[Address, Port | _] | _]] | T], Index, Acc) ->
+    SlotsMap =
+        #slots_map{
+            index = Index,
+            start_slot = binary_to_integer(StartSlot),
+            end_slot = binary_to_integer(EndSlot),
+            node = #node{
+                address = binary_to_list(Address),
+                port = binary_to_integer(Port)
+            }
+        },
+    parse_cluster_slots(T, Index+1, [SlotsMap | Acc]);
+parse_cluster_slots([], _Index, Acc) ->
+    lists:reverse(Acc).
+
+
+-spec connect_all_slots(ClusterName::atom(), [#slots_map{}]) -> [integer()].
+connect_all_slots(ClusterName, SlotsMapList) ->
+    [SlotsMap#slots_map{node=connect_node(ClusterName, SlotsMap#slots_map.node)} ||
+     SlotsMap <- SlotsMapList].
+
+
+-spec connect_node(ClusterName::atom(), #node{}) -> #node{} | undefined.
+connect_node(ClusterName, Node) ->
+    Res = lookup_eredis_pid(ClusterName, Node#node.address, Node#node.port),
+    case Res of
+        {ok, _} -> Node;
+        _ -> undefined
+    end.
+
+
+%% Creates tuple with one element per Redis slot. Each element maps the Redis slot to the Redis
+%% node index in SlotsMaps.
+-spec create_slots_cache([#slots_map{}]) -> [integer()].
+create_slots_cache(SlotsMaps) ->
+  SlotsCache = [[{Index, SlotsMap#slots_map.index}
+        || Index <- lists:seq(SlotsMap#slots_map.start_slot, SlotsMap#slots_map.end_slot)]
+        || SlotsMap <- SlotsMaps],
+  FlatSlotsCache = lists:flatten(SlotsCache),
+  SortedSlotsCache = lists:sort(FlatSlotsCache),
+  [ Index || {_,Index} <- SortedSlotsCache].
+
+
+-spec cache_eredis_pids(State::#state{}, Slot::integer()) -> ok.
+cache_eredis_pids(State, Slot) ->
+    RedisNodeIndex = element(Slot + 1, State#state.slots_cache),
+    SlotsMap = element(RedisNodeIndex, State#state.slots_maps),
+    NewPid = if
+        SlotsMap#slots_map.node =/= undefined ->
+            {ok, Pid} = lookup_eredis_pid(State#state.cluster_name,
+                                          SlotsMap#slots_map.node#node.address,
+                                          SlotsMap#slots_map.node#node.port),
+            Pid;
+        true ->
+            undefined
+    end,
+    ets:insert(ets_table_name(State#state.cluster_name), {Slot, {NewPid, State#state.version}}),
```

Comment:
I think it is confusing to use one ets table for both {Host, Post} -> Pid map and also fro the Slot -> {Pid, version}.
I would have put this into 2 different tables. Also maybe we only need the Slot -> {Pid, version} to be a ets table. Since the {Host, Post -> Pid map is basically only accessed from the gen_server process... I'm fine with 2 ets tables...

Reply:
maybe rename the tables to be: `cluster_name_node_pids` and `cluster_name_slot_pids`!